### ISSUE TITLE
Updated to latest update fom 3/21/2023, and I don't see any of my groups in Panorama View

### ISSUE THREAD

**Issue by roeyk** [photodiode/panorama-view#141]
I still see the tabs when I click on the down-arrow on the right of the Firefox menu bar, but when I switch to Panorama view, I get an empty group where there should be three or four groups full of tabs.

**Comment by photodiode** [comment]
If you open the console in the panorama view and reload the page, does any errors appear?
I'm really curious as to what could have caused this.

**Comment by roeyk** [comment]
Good question.  How do I open the console in Panorama?  Now I see one group with all the tabs squished into it.  Where does this extension save its group data to, which file? I can pull another one from backup if need be, and compare how the extension behaves with the old file.

**Comment by photodiode** [comment]
For console: Open Firefox menu -> More tools -> Browser Console

The data is stored in the window and tab's session storage, so I would guess the data is in the file "**sessionstore.jsonlz4**" in the Firefox profile folder.

It sounds to me like the session storage got erased somehow.

**Comment by roeyk** [comment]
So I checked the console, didn't see anything specifically when i re-loaded this page, other than "[NoScript]:0 Prompt Hook installation https://github.com/photodiode/panorama-view/issues/141".   I tried re-loading with the sessionstore.jsonl4z from a known good backup, and get the same issue.   So I think it's the upgrade to the new version of the tab groups script that maybe did it.

**Comment by photodiode** [comment]
It happened to me for the first time ever this morning. All tabs in one group.
The frustrating part is I have no idea why all tabs and windows would suddenly lose their group value.
I did upgrade Firefox to v111.0 yesterday, maybe that is what caused this..? Seems weird.

**Comment by roeyk** [comment]
Yeah it seems to be a repeatable issue here for me.  I exit out of Firefox, restart my system, and all the tab groups have disappeared.

**Comment by roeyk** [comment]
Also, a bunch of tabs have somehow become "hidden" tabs.

**Comment by harry-g** [comment]
Same problem here. 
Firefox 111.01 on Mac.
Lost all tab groups 2 times already.

But I don't have the Firefox extension "Select After Closing Current".

To me it seems this happend during a reboot and not when I closed the browser myself. 
But just a guess.
Or it could be when I had several Windows open, which was formerly no problem by reopening the one with tab groups with "recently closed windows".

**Comment by photodiode** [comment]
The only things I can think of that could suddenly cause this are:
1. Sometimes Firefox closes before the saveGroups function could finish (I find that weird)
2. Something changed so that a race condition with the group cache became more frequent

I would assume it's number 2 and, I don't know why it didn't occur to be before but, I should have added a mutex lock on the group cache object and save function.
The cache is obviously being used asynchronously from different functions.

**Comment by photodiode** [comment]
I will try to get a new version out next week.

I really wish Firefox would hurry up and get itself a standard tab groups API soon so I didn't have to do all this hacky emulation anymore.. :'(
